File: backend/idea_studio.py
from __future__ import annotations

import asyncio
import json
from datetime import datetime, timezone
from typing import Any, Awaitable, Callable

ModelCaller = Callable[[str, str, float], Awaitable[str]]


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _parse_json(text: str) -> Any:
    value = text.strip()
    if value.startswith("```"):
        lines = value.splitlines()
        if lines and lines[0].startswith("```"):
            lines = lines[1:]
        if lines and lines[-1].strip() == "```":
            lines = lines[:-1]
        value = "\n".join(lines)
    return json.loads(value)


async def _model_json(
    chat: ModelCaller,
    system: str,
    payload: Any,
    *,
    temperature: float = 0.2,
) -> Any:
    raw = await chat(system, json.dumps(payload, ensure_ascii=False), temperature)
    return _parse_json(raw)


def _candidate(candidate: dict[str, Any], index: int) -> dict[str, Any]:
    now = _now()
    return {
        "id": str(candidate.get("id") or f"idea-{int(datetime.now().timestamp() * 1000)}-{index}"),
        "title": str(candidate.get("title") or "Untitled research candidate"),
        "problem": str(candidate.get("problem") or "Problem still needs to be explicit."),
        "gap": str(candidate.get("gap") or "Gap requires literature verification."),
        "mechanism": str(candidate.get("mechanism") or "Mechanism not yet specified."),
        "hypothesis": str(candidate.get("hypothesis") or "Hypothesis not yet falsifiable."),
        "noveltyClaim": str(
            candidate.get("noveltyClaim")
            or "Novelty claim is provisional pending closest-work checks."
        ),
        "resourceFit": str(candidate.get("resourceFit") or "Resource fit has not been checked."),
        "experiment": str(
            candidate.get("experiment")
            or "Define a minimum discriminating experiment before scaling."
        ),
        "noveltyRisk": str(
            candidate.get("noveltyRisk") or "Unknown until closest prior work is checked."
        ),
        "origin": str(candidate.get("origin") or "ARIS-style discovery"),
        "version": int(candidate.get("version") or 1),
        "status": str(candidate.get("status") or "candidate"),
        "createdAt": str(candidate.get("createdAt") or now),
    }


def _critique(critique: dict[str, Any], candidate_id: str, index: int) -> dict[str, Any]:
    severity = str(critique.get("severity") or "major")
    if severity not in {"blocker", "major", "minor"}:
        severity = "major"
    return {
        "id": str(critique.get("id") or f"{candidate_id}-critic-{index}"),
        "candidateId": candidate_id,
        "role": str(critique.get("role") or "Research critic"),
        "source": str(critique.get("source") or "Idea Studio"),
        "summary": str(critique.get("summary") or ""),
        "objections": [str(x) for x in critique.get("objections", [])],
        "evidenceNeeded": [str(x) for x in critique.get("evidenceNeeded", [])],
        "discriminatingExperiment": str(critique.get("discriminatingExperiment") or ""),
        "severity": severity,
    }


def _evidence_digest(items: list[dict[str, Any]] | None) -> list[dict[str, Any]]:
    output: list[dict[str, Any]] = []
    for item in (items or [])[:30]:
        output.append(
            {
                "title": item.get("title"),
                "stance": item.get("stance"),
                "linkedClaim": item.get("linkedClaim"),
                "note": item.get("note"),
                "year": item.get("year"),
                "venue": item.get("venue"),
            }
        )
    return output


def _finding_digest(items: list[dict[str, Any]] | None) -> list[dict[str, Any]]:
    output: list[dict[str, Any]] = []
    for item in (items or [])[:30]:
        output.append(
            {
                "kind": item.get("kind"),
                "statement": item.get("statement"),
                "evidence": item.get("evidence"),
                "conditions": item.get("conditions"),
                "source": item.get("source"),
            }
        )
    return output


async def discover(payload: dict[str, Any], chat: ModelCaller) -> dict[str, Any]:
    system = """
You are the Idea Discovery core inside PaperForge Idea Studio.

Follow an ARIS-style research discovery discipline:
1. Start from the user's real assets, constraints, prior work, and observed failures.
2. Diverge before converging. Produce candidates from materially different discovery routes:
   - data/asset-driven opportunity,
   - prior-art limitation or assumption,
   - failure/negative-result driven opportunity,
   - cross-domain mechanism transfer only when the prerequisites plausibly match.
3. A candidate is a hypothesis to investigate, not a novelty claim.
4. Never say "nobody has studied this" unless evidence supports that statement.
5. Novelty must be checked against direct competitors, adjacent methods, negative results, and analogous mechanisms.
6. Prefer a minimum discriminating experiment over a large implementation.
7. If a critical user fact is missing, ask a focused question instead of inventing the fact.
8. Use prior Findings as constraints; do not repeat a failed route without a reason it may now work.

Return STRICT JSON with:
{
  "assistantMessage": string,
  "briefPatch": object,
  "questions": string[],
  "searchQueries": string[],
  "candidates": [
    {
      "id": string,
      "title": string,
      "problem": string,
      "gap": string,
      "mechanism": string,
      "hypothesis": string,
      "noveltyClaim": string,
      "resourceFit": string,
      "experiment": string,
      "noveltyRisk": string,
      "origin": string,
      "version": 1,
      "status": "candidate" | "needs-evidence"
    }
  ],
  "controller": {
    "action": "ask-user" | "search-more" | "revise" | "pilot",
    "rationale": string,
    "question": string,
    "searchQueries": string[],
    "experiment": string,
    "stopCondition": string
  }
}

Generate at most 4 candidates. Keep every novelty statement explicitly provisional.
""".strip()

    data = {
        "venue": payload.get("venue"),
        "researchBrief": payload.get("researchBrief") or {},
        "currentIdeaSpec": payload.get("ideaSpec") or {},
        "userInstruction": payload.get("instruction"),
        "conversationTail": (payload.get("messages") or [])[-8:],
        "existingCandidates": (payload.get("candidates") or [])[-8:],
        "evidence": _evidence_digest(payload.get("evidence")),
        "findings": _finding_digest(payload.get("findings")),
    }
    result = await _model_json(chat, system, data, temperature=0.45)
    candidates = [
        _candidate(candidate, index)
        for index, candidate in enumerate(result.get("candidates") or [])
        if isinstance(candidate, dict)
    ]
    controller = result.get("controller") if isinstance(result.get("controller"), dict) else None
    if controller:
        controller = {
            "candidateId": controller.get("candidateId"),
            "action": controller.get("action") or "search-more",
            "rationale": controller.get("rationale") or "More evidence is needed.",
            "question": controller.get("question") or "",
            "searchQueries": [str(x) for x in controller.get("searchQueries", [])],
            "experiment": controller.get("experiment") or "",
            "stopCondition": controller.get("stopCondition") or "",
        }
    return {
        "assistantMessage": str(
            result.get("assistantMessage")
            or "Candidate discovery completed. Treat every route as provisional until evidence and a discriminating test support it."
        ),
        "briefPatch": result.get("briefPatch") if isinstance(result.get("briefPatch"), dict) else {},
        "questions": [str(x) for x in result.get("questions", [])],
        "searchQueries": [str(x) for x in result.get("searchQueries", [])],
        "candidates": candidates,
        "controller": controller,
    }


CRITIC_ROLES = [
    (
        "Problem & Value Reviewer",
        "Agent Laboratory-style specialist role",
        "Judge whether the stated problem is real, measurable, and important enough to motivate a research contribution. Separate problem importance from method attractiveness.",
    ),
    (
        "Novelty & Evidence Reviewer",
        "PaperQA2 / InnoEval-style evidence audit",
        "Audit claim-to-evidence alignment. Identify the closest work that must be checked, distinguish missing evidence from evidence of absence, and attack recombination-only novelty.",
    ),
    (
        "Method & Mechanism Reviewer",
        "Agent Laboratory-style method specialist",
        "Test whether the mechanism plausibly explains the expected effect, whether simpler alternatives exist, and which ablation would distinguish them.",
    ),
    (
        "Data & Experiment Reviewer",
        "AI-Scientist-v2-style experiment gate",
        "Check whether the available resources can actually test the hypothesis. Propose the cheapest discriminating pilot and an explicit stop condition.",
    ),
]


async def critic(payload: dict[str, Any], chat: ModelCaller) -> dict[str, Any]:
    candidate = payload.get("candidate") or {}
    candidate_id = str(candidate.get("id") or "candidate")

    async def one(role: str, source: str, instruction: str) -> dict[str, Any]:
        system = f"""
You are {role} inside a research Critic Council.
Method reference: {source}.
{instruction}

Review independently. Do not optimize for agreement and do not predict paper acceptance.
Return STRICT JSON with:
{{
  "role": "{role}",
  "source": "{source}",
  "summary": string,
  "objections": string[],
  "evidenceNeeded": string[],
  "discriminatingExperiment": string,
  "severity": "blocker" | "major" | "minor"
}}
A blocker means the current claim cannot responsibly proceed without resolving it.
""".strip()
        data = {
            "venue": payload.get("venue"),
            "researchBrief": payload.get("researchBrief") or {},
            "candidate": candidate,
            "evidence": _evidence_digest(payload.get("evidence")),
            "findings": _finding_digest(payload.get("findings")),
        }
        result = await _model_json(chat, system, data, temperature=0.15)
        return result if isinstance(result, dict) else {}

    raw = await asyncio.gather(
        *(one(role, source, instruction) for role, source, instruction in CRITIC_ROLES)
    )
    return {
        "critiques": [
            _critique(item, candidate_id, index)
            for index, item in enumerate(raw)
        ]
    }


async def decide(payload: dict[str, Any], chat: ModelCaller) -> dict[str, Any]:
    system = """
You are the Research Controller for PaperForge Idea Studio.

You do NOT write a paper and you do NOT reward longer workflows.
Choose the next action that is most likely to change the current research decision at the lowest reasonable cost.

Allowed actions:
- ask-user: a missing human fact blocks responsible progress.
- search-more: prior-art or evidence uncertainty is the dominant blocker.
- revise: the route is promising but the problem/mechanism/hypothesis must change first.
- pilot: a small bounded experiment can discriminate the key hypothesis.
- adopt: the route is sufficiently specified for an IdeaSpec; this does not mean the paper is proven.
- stop: evidence or feasibility makes the route not worth continued work now.

Rules:
- Never choose adopt merely because reviewers agree.
- A blocker critique must be addressed or explicitly bounded.
- If novelty evidence is weak, search before scaling experiments.
- If a pilot is chosen, include a stop condition.
- If user resources cannot support the experiment, ask or revise instead of pretending they can.

Return STRICT JSON:
{
  "decision": {
    "candidateId": string,
    "action": "ask-user" | "search-more" | "revise" | "pilot" | "adopt" | "stop",
    "rationale": string,
    "question": string,
    "searchQueries": string[],
    "experiment": string,
    "stopCondition": string
  }
}
""".strip()
    data = {
        "venue": payload.get("venue"),
        "researchBrief": payload.get("researchBrief") or {},
        "candidate": payload.get("candidate") or {},
        "critiques": payload.get("critiques") or [],
        "evidence": _evidence_digest(payload.get("evidence")),
        "findings": _finding_digest(payload.get("findings")),
    }
    result = await _model_json(chat, system, data, temperature=0.05)
    decision = result.get("decision") if isinstance(result, dict) else None
    if not isinstance(decision, dict):
        raise ValueError("Research Controller returned no decision object")
    return {
        "decision": {
            "candidateId": decision.get("candidateId")
            or (payload.get("candidate") or {}).get("id"),
            "action": decision.get("action") or "revise",
            "rationale": str(decision.get("rationale") or ""),
            "question": str(decision.get("question") or ""),
            "searchQueries": [str(x) for x in decision.get("searchQueries", [])],
            "experiment": str(decision.get("experiment") or ""),
            "stopCondition": str(decision.get("stopCondition") or ""),
        }
    }


async def refine(payload: dict[str, Any], chat: ModelCaller) -> dict[str, Any]:
    system = """
You are the refinement stage of PaperForge Idea Studio.

Revise the candidate only where evidence, critiques, user constraints, or Findings justify a change.
Do not hide unresolved objections by rewriting the title.
Preserve the same stable candidate id, increment version by exactly 1, and keep novelty language provisional.
If a critique cannot be resolved with current evidence, make the uncertainty explicit in noveltyRisk.

Return STRICT JSON:
{
  "assistantMessage": string,
  "candidate": {
    "id": string,
    "title": string,
    "problem": string,
    "gap": string,
    "mechanism": string,
    "hypothesis": string,
    "noveltyClaim": string,
    "resourceFit": string,
    "experiment": string,
    "noveltyRisk": string,
    "origin": string,
    "version": integer,
    "status": "candidate" | "needs-evidence" | "pilot-ready" | "paused"
  }
}
""".strip()
    current = payload.get("candidate") or {}
    data = {
        "venue": payload.get("venue"),
        "researchBrief": payload.get("researchBrief") or {},
        "candidate": current,
        "critiques": payload.get("critiques") or [],
        "evidence": _evidence_digest(payload.get("evidence")),
        "findings": _finding_digest(payload.get("findings")),
    }
    result = await _model_json(chat, system, data, temperature=0.2)
    candidate = result.get("candidate") if isinstance(result, dict) else None
    if not isinstance(candidate, dict):
        raise ValueError("Refinement returned no candidate")
    candidate["id"] = current.get("id") or candidate.get("id")
    candidate["version"] = int(current.get("version") or 1) + 1
    candidate["createdAt"] = current.get("createdAt") or _now()
    return {
        "assistantMessage": str(result.get("assistantMessage") or "Candidate refined."),
        "candidate": _candidate(candidate, 0),
    }


async def handle_idea_studio(
    workflow: str,
    payload: dict[str, Any],
    chat: ModelCaller,
) -> dict[str, Any]:
    if workflow == "idea.discover":
        return await discover(payload, chat)
    if workflow == "idea.critic":
        return await critic(payload, chat)
    if workflow == "idea.decide":
        return await decide(payload, chat)
    if workflow == "idea.refine":
        return await refine(payload, chat)
    raise ValueError(f"Unsupported Idea Studio workflow: {workflow}")

File: src/lib/ideaStudio.ts
import type { EvidenceItem, IdeaSpec } from './research'

export type IdeaCandidateStatus = 'candidate' | 'needs-evidence' | 'pilot-ready' | 'adopted' | 'paused' | 'rejected'
export type ControllerAction = 'ask-user' | 'search-more' | 'revise' | 'pilot' | 'adopt' | 'stop'
export type FindingKind = 'observation' | 'success' | 'failure' | 'constraint'

export interface ResearchBrief {
  researchGoal: string
  assets: string
  constraints: string
  priorWork: string
  targetOutcome: string
  updatedAt?: string
}

export interface StudioMessage {
  id: string
  role: 'user' | 'assistant'
  content: string
  createdAt: string
}

export interface IdeaCandidate {
  id: string
  title: string
  problem: string
  gap: string
  mechanism: string
  hypothesis: string
  noveltyClaim: string
  resourceFit: string
  experiment: string
  noveltyRisk: string
  origin: string
  version: number
  status: IdeaCandidateStatus
  createdAt: string
}

export interface IdeaCritique {
  id: string
  candidateId: string
  role: string
  source: string
  summary: string
  objections: string[]
  evidenceNeeded: string[]
  discriminatingExperiment: string
  severity: 'blocker' | 'major' | 'minor'
}

export interface ResearchFinding {
  id: string
  kind: FindingKind
  statement: string
  evidence: string
  conditions: string
  source: string
  createdAt: string
}

export interface ControllerDecision {
  candidateId?: string
  action: ControllerAction
  rationale: string
  question?: string
  searchQueries?: string[]
  experiment?: string
  stopCondition?: string
}

export interface IdeaStudioState {
  brief: ResearchBrief
  messages: StudioMessage[]
  candidates: IdeaCandidate[]
  critiques: IdeaCritique[]
  findings: ResearchFinding[]
  searchQueries: string[]
  decision?: ControllerDecision
}

export interface IdeaDiscoveryResult {
  assistantMessage: string
  briefPatch?: Partial<ResearchBrief>
  questions?: string[]
  searchQueries?: string[]
  candidates?: IdeaCandidate[]
  controller?: ControllerDecision
}

export interface IdeaCritiqueResult {
  critiques: IdeaCritique[]
}

export interface IdeaDecisionResult {
  decision: ControllerDecision
}

export interface IdeaRefineResult {
  candidate: IdeaCandidate
  assistantMessage?: string
}

export const emptyResearchBrief: ResearchBrief = {
  researchGoal: '',
  assets: '',
  constraints: '',
  priorWork: '',
  targetOutcome: '',
}

export const emptyIdeaStudioState: IdeaStudioState = {
  brief: { ...emptyResearchBrief },
  messages: [],
  candidates: [],
  critiques: [],
  findings: [],
  searchQueries: [],
}

export function normalizeIdeaStudioState(input: unknown): IdeaStudioState {
  const value = input && typeof input === 'object' ? input as Partial<IdeaStudioState> : {}
  const brief = value.brief && typeof value.brief === 'object'
    ? { ...emptyResearchBrief, ...value.brief }
    : { ...emptyResearchBrief }

  return {
    ...emptyIdeaStudioState,
    ...value,
    brief,
    messages: Array.isArray(value.messages) ? value.messages : [],
    candidates: Array.isArray(value.candidates)
      ? value.candidates.map((candidate, index) => normalizeCandidate(candidate, index))
      : [],
    critiques: Array.isArray(value.critiques) ? value.critiques : [],
    findings: Array.isArray(value.findings) ? value.findings : [],
    searchQueries: Array.isArray(value.searchQueries) ? value.searchQueries : [],
  }
}

export function nowIso() {
  return new Date().toISOString()
}

export function makeMessage(role: StudioMessage['role'], content: string): StudioMessage {
  return {
    id: role + '-' + Date.now() + '-' + Math.random().toString(36).slice(2, 7),
    role,
    content,
    createdAt: nowIso(),
  }
}

export function normalizeCandidate(candidate: Partial<IdeaCandidate>, index = 0): IdeaCandidate {
  const createdAt = candidate.createdAt || nowIso()
  return {
    id: candidate.id || 'idea-' + Date.now() + '-' + index,
    title: candidate.title || 'Untitled research candidate',
    problem: candidate.problem || 'Problem still needs to be made explicit.',
    gap: candidate.gap || 'Gap requires literature verification.',
    mechanism: candidate.mechanism || 'Mechanism not yet specified.',
    hypothesis: candidate.hypothesis || 'Hypothesis not yet falsifiable.',
    noveltyClaim: candidate.noveltyClaim || 'Novelty claim is provisional pending prior-art checks.',
    resourceFit: candidate.resourceFit || 'Resource fit has not been checked.',
    experiment: candidate.experiment || 'Define a minimum discriminating experiment before investment.',
    noveltyRisk: candidate.noveltyRisk || 'Unknown until closest prior work is checked.',
    origin: candidate.origin || 'Idea Studio',
    version: candidate.version || 1,
    status: candidate.status || 'candidate',
    createdAt,
  }
}

export function mergeCandidates(current: IdeaCandidate[], incoming: IdeaCandidate[]) {
  const map = new Map(current.map((candidate) => [candidate.id, candidate]))
  for (const candidate of incoming) {
    map.set(candidate.id, normalizeCandidate(candidate))
  }
  return [...map.values()]
}

export function candidateToIdeaSpec(candidate: IdeaCandidate, current: IdeaSpec): IdeaSpec {
  return {
    ...current,
    title: candidate.title,
    topic: current.topic || candidate.title,
    problem: candidate.problem,
    gap: candidate.gap,
    coreIdea: candidate.mechanism,
    hypothesis: candidate.hypothesis,
    noveltyClaim: candidate.noveltyClaim,
    experiments: candidate.experiment,
    datasets: current.datasets || candidate.resourceFit,
    risks: [current.risks, 'Novelty / positioning risk: ' + candidate.noveltyRisk].filter(Boolean).join('\n'),
  }
}

export function offlineDiscover(params: {
  brief: ResearchBrief
  currentIdea: IdeaSpec
  evidence: EvidenceItem[]
  userMessage: string
  findings: ResearchFinding[]
}): IdeaDiscoveryResult {
  const goal = params.brief.researchGoal.trim() || params.currentIdea.topic.trim() || params.userMessage.trim()
  const assets = params.brief.assets.trim() || params.currentIdea.datasets.trim()
  const constraints = params.brief.constraints.trim()
  if (!goal) {
    return {
      assistantMessage:
        'Before generating candidates, describe the research direction or problem you want to investigate. I will keep the request narrow and use your available data and constraints as hard inputs.',
      controller: {
        action: 'ask-user',
        rationale: 'The research goal is still unspecified.',
        question: 'What problem or research direction do you want to investigate?',
      },
    }
  }

  const evidenceNote = params.evidence.length
    ? params.evidence.length + ' evidence items are already available and should be checked before strengthening novelty claims.'
    : 'No literature evidence is attached yet, so every gap and novelty statement below is provisional.'
  const findingNote = params.findings.length
    ? 'Existing findings should be used to avoid repeating failed routes.'
    : 'No prior experiment findings have been recorded yet.'

  const candidates = [
    normalizeCandidate({
      id: 'data-opportunity-' + Date.now(),
      title: 'Data-driven opportunity: ' + goal,
      problem: 'Identify a recurring failure, bottleneck, or unexplained structure directly observable in the available research assets.',
      gap: 'Determine whether existing methods explicitly model this observed structure under comparable assumptions.',
      mechanism: 'Exploit a property that is actually present in the available data instead of starting from a method name.',
      hypothesis: 'If the observed structure matters, a method that models it explicitly should outperform or explain a strong simpler baseline under a controlled comparison.',
      noveltyClaim: 'Provisional: the contribution would be the explicit modeling and validation of a data-observed mechanism, not merely applying an LLM or agent.',
      resourceFit: assets || 'Available assets still need to be listed and verified.',
      experiment: 'Start with a small descriptive analysis and one discriminating baseline comparison before building a complex system.',
      noveltyRisk: evidenceNote,
      origin: 'ARIS-style data-driven discovery',
      status: params.evidence.length ? 'candidate' : 'needs-evidence',
    }, 0),
    normalizeCandidate({
      id: 'gap-opportunity-' + (Date.now() + 1),
      title: 'Evidence-gap opportunity: ' + goal,
      problem: 'Find a concrete limitation or assumption shared by the closest existing approaches.',
      gap: 'A valid gap must survive direct-competitor, adjacent-field, negative-result, and contradictory-evidence searches.',
      mechanism: 'Change one specific assumption, representation, or decision step that explains the limitation.',
      hypothesis: 'If the targeted assumption is the true bottleneck, changing it should improve a predefined outcome without requiring unrelated changes.',
      noveltyClaim: 'Provisional: novelty depends on the closest prior work and must be phrased as the smallest defensible difference.',
      resourceFit: assets || 'Resource fit requires verification.',
      experiment: 'Compare against the strongest fair prior method while holding data and evaluation constant.',
      noveltyRisk: 'High until closest prior work is retrieved. ' + evidenceNote,
      origin: 'ARIS-style literature-gap discovery',
      status: 'needs-evidence',
    }, 1),
    normalizeCandidate({
      id: 'failure-opportunity-' + (Date.now() + 2),
      title: 'Failure-driven opportunity: ' + goal,
      problem: 'Use failed or degraded cases as research signal rather than hiding them.',
      gap: 'Check whether existing work explains, predicts, or repairs the failure mode under the same constraints.',
      mechanism: 'Convert the failure evidence into a constraint, feedback signal, or adaptive decision rule.',
      hypothesis: 'If the failure mode is systematic, incorporating its evidence should reduce that failure without simply rejecting all difficult cases.',
      noveltyClaim: 'Provisional: the contribution would be a failure-grounded mechanism plus evidence that it changes the relevant behavior.',
      resourceFit: assets || 'Requires at least examples of failures, edge cases, or rejected routes.',
      experiment: 'Build a compact failure set, test the simplest corrective mechanism, and define what result would falsify the explanation.',
      noveltyRisk: findingNote + ' ' + evidenceNote,
      origin: 'ARIS-style failure-driven discovery',
      status: 'candidate',
    }, 2),
  ]

  return {
    assistantMessage:
      'I generated three deliberately different research routes: one starts from your data, one from the prior-art gap, and one from failures. They are candidates, not claims of novelty. The next useful step is to pick one route or add evidence that can eliminate a route.',
    candidates,
    searchQueries: [
      goal + ' closest prior work',
      goal + ' limitations failure cases',
      goal + ' benchmark baseline',
    ],
    controller: {
      action: params.evidence.length ? 'revise' : 'search-more',
      rationale: params.evidence.length
        ? 'There is enough initial context to discuss and refine candidates, but novelty still needs explicit closest-work comparison.'
        : 'The candidates are under-grounded because no literature evidence has been attached.',
      searchQueries: [
        goal + ' closest prior work',
        goal + ' limitations failure cases',
      ],
    },
    briefPatch: {
      researchGoal: params.brief.researchGoal || goal,
      updatedAt: nowIso(),
      ...(constraints ? { constraints } : {}),
    },
  }
}

export function offlineCritiques(candidate: IdeaCandidate, evidence: EvidenceItem[]): IdeaCritique[] {
  const evidenceMissing = evidence.length === 0
  const roles = [
    {
      role: 'Problem & Value Reviewer',
      source: 'Agent Laboratory-style specialist role',
      summary: 'Check whether the problem changes an important scientific or engineering decision.',
      objections: ['The candidate may describe a method before proving that the underlying problem is important.'],
      evidenceNeeded: ['A concrete failure, cost, limitation, or unmet capability that can be measured.'],
      discriminatingExperiment: 'Show that the problem exists on a representative sample before testing the proposed mechanism.',
      severity: 'major' as const,
    },
    {
      role: 'Novelty & Evidence Reviewer',
      source: 'PaperQA2 / InnoEval-style evidence audit',
      summary: evidenceMissing
        ? 'The novelty claim is not grounded because no evidence is attached.'
        : 'Evidence exists, but the closest-work comparison still has to support the exact claimed difference.',
      objections: ['A renamed combination of known components is not automatically a research contribution.'],
      evidenceNeeded: ['Closest direct competitor', 'Adjacent-field analogue', 'Negative or contradictory evidence'],
      discriminatingExperiment: 'Not an experiment first: retrieve the closest work and write a claim-level difference table.',
      severity: evidenceMissing ? 'blocker' as const : 'major' as const,
    },
    {
      role: 'Method & Mechanism Reviewer',
      source: 'Agent Laboratory-style method specialist',
      summary: 'The mechanism must explain why the proposed change should affect the target outcome.',
      objections: ['A correlation between the method and metric would not by itself validate the proposed mechanism.'],
      evidenceNeeded: ['Ablation or controlled comparison tied to the mechanism.'],
      discriminatingExperiment: candidate.experiment,
      severity: 'major' as const,
    },
    {
      role: 'Data & Experiment Reviewer',
      source: 'AI-Scientist-v2-style experimental gate',
      summary: 'Start with the cheapest experiment that can change belief in the candidate.',
      objections: ['The candidate may require data, labels, or execution access that are not currently available.'],
      evidenceNeeded: ['Verified data fields, sample count, labels or executable environment, and evaluation metric.'],
      discriminatingExperiment: 'Run a bounded pilot with an explicit stop condition before scaling.',
      severity: 'major' as const,
    },
  ]
  return roles.map((role, index) => ({
    id: candidate.id + '-critic-' + index,
    candidateId: candidate.id,
    ...role,
  }))
}

export function offlineDecision(params: {
  candidate: IdeaCandidate
  critiques: IdeaCritique[]
  evidence: EvidenceItem[]
  findings: ResearchFinding[]
}): ControllerDecision {
  const blockers = params.critiques.filter((critique) => critique.severity === 'blocker')
  if (blockers.length) {
    return {
      candidateId: params.candidate.id,
      action: 'search-more',
      rationale: 'At least one blocker is unresolved. The next action should reduce uncertainty rather than expand the method.',
      searchQueries: ['closest prior work for ' + params.candidate.title],
      stopCondition: 'Stop or revise this route if the closest work already implements the same mechanism under comparable assumptions.',
    }
  }
  if (params.evidence.length < 3) {
    return {
      candidateId: params.candidate.id,
      action: 'search-more',
      rationale: 'The candidate has too little evidence for a strong novelty or gap claim.',
      searchQueries: ['closest prior work for ' + params.candidate.title, params.candidate.title + ' limitations'],
    }
  }
  if (!params.findings.length) {
    return {
      candidateId: params.candidate.id,
      action: 'pilot',
      rationale: 'The candidate has enough context for a bounded discriminating experiment, but no execution evidence has been recorded.',
      experiment: params.candidate.experiment,
      stopCondition: 'Do not scale the project if the pilot cannot distinguish the proposed mechanism from a strong simpler baseline.',
    }
  }
  return {
    candidateId: params.candidate.id,
    action: 'revise',
    rationale: 'Use the accumulated findings and critiques to refine the mechanism before adoption.',
  }
}
